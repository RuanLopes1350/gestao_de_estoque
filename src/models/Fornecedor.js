import mongoose from 'mongoose';
import mongoosePaginate from 'mongoose-paginate-v2';

class Fornecedor {
    constructor() {
        const fornecedorSchema = new mongoose.Schema(
            {
                nome_fornecedor: {type: String, required: true, index: true},
                cnpj: {type: String, required: true, index: true},
                endereco: [
                    {
                        telefone: {type: String, required: true},
                        email: {type: String, required: true},
                        
                    }
                ]
            },
            {
                timestamps: {createdAt: 'data_cadastro', updatedAt: 'data_ultima_atualizacao'},
                versionKey: false,
            }
        )
        fornecedorSchema.plugin(mongoosePaginate);
        this.model = mongoose.model('fornecedores', fornecedorSchema);
    }
}

export default new Fornecedor().model;