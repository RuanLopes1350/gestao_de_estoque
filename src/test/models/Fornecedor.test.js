import mongoose from 'mongoose';
import Fornecedor from '../../models/Fornecedor';
import { MongoMemoryServer } from 'mongodb-memory-server';

let mongoServer;

beforeAll(async () => {
  mongoServer = await MongoMemoryServer.create();
  const uri = mongoServer.getUri();
  await mongoose.connect(uri);
});

afterAll(async () => {
  await mongoose.disconnect();
  await mongoServer.stop();
});

afterEach(async () => {
  jest.clearAllMocks();
  await Fornecedor.deleteMany({});
});

describe('Modelo de Fornecedor', () => {
  it('deve criar um fornecedor com dados válidos', async () => {
    const fornecedorData = {
      nome_fornecedor: 'Fornecedor Teste',
      cnpj: '12345678000199',
      telefone: '11999999999',
      email: 'fornecedor@exemplo.com',
      endereco: [
        {
          logradouro: 'Rua Exemplo',
          bairro: 'Centro',
          cidade: 'São Paulo',
          estado: 'SP',
          cep: '01000-000',
        },
      ],
    };

    const fornecedor = new Fornecedor(fornecedorData);
    await fornecedor.save();

    const saved = await Fornecedor.findById(fornecedor._id);
    expect(saved.nome_fornecedor).toBe(fornecedorData.nome_fornecedor);
    expect(saved.cnpj).toBe(fornecedorData.cnpj);
    expect(saved.telefone).toBe(fornecedorData.telefone);
    expect(saved.email).toBe(fornecedorData.email);
    expect(saved.endereco).toHaveLength(1);
    expect(saved.endereco[0].cidade).toBe('São Paulo');
  });

  it('deve falhar ao criar um fornecedor sem nome', async () => {
    const fornecedorData = {
      cnpj: '12345678000199',
      telefone: '11999999999',
      email: 'fornecedor@exemplo.com',
      endereco: [
        {
          logradouro: 'Rua Exemplo',
          bairro: 'Centro',
          cidade: 'São Paulo',
          estado: 'SP',
          cep: '01000-000',
        },
      ],
    };

    const fornecedor = new Fornecedor(fornecedorData);
    await expect(fornecedor.save()).rejects.toThrowErrorMatchingSnapshot();
  });

  it('deve falhar ao criar um fornecedor sem CNPJ', async () => {
    const fornecedorData = {
      nome_fornecedor: 'Fornecedor Teste',
      telefone: '11999999999',
      email: 'fornecedor@exemplo.com',
      endereco: [
        {
          logradouro: 'Rua Exemplo',
          bairro: 'Centro',
          cidade: 'São Paulo',
          estado: 'SP',
          cep: '01000-000',
        },
      ],
    };

    const fornecedor = new Fornecedor(fornecedorData);
    await expect(fornecedor.save()).rejects.toThrowErrorMatchingSnapshot();
  });

  it('deve falhar se algum campo do endereço estiver ausente', async () => {
    const fornecedorData = {
      nome_fornecedor: 'Fornecedor Teste',
      cnpj: '12345678000199',
      telefone: '11999999999',
      email: 'fornecedor@exemplo.com',
      endereco: [
        {
          logradouro: 'Rua Exemplo',
          bairro: 'Centro',
          cidade: 'São Paulo',
          estado: 'SP',
        },
      ],
    };

    const fornecedor = new Fornecedor(fornecedorData);
    await expect(fornecedor.save()).rejects.toThrowErrorMatchingSnapshot();
  });
});
