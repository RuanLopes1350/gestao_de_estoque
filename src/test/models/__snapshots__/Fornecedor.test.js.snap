// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modelo de Fornecedor deve falhar ao criar um fornecedor sem CNPJ 1`] = `"fornecedores validation failed: cnpj: Path \`cnpj\` is required."`;

exports[`Modelo de Fornecedor deve falhar ao criar um fornecedor sem nome 1`] = `"fornecedores validation failed: nome_fornecedor: Path \`nome_fornecedor\` is required."`;

exports[`Modelo de Fornecedor deve falhar se algum campo do endere√ßo estiver ausente 1`] = `"fornecedores validation failed: endereco.0.cep: Path \`cep\` is required."`;
